[WIP] PTB semantics doc

tnowacki opened the pull request:
## Description 

- WIP for several docs. My layout doesn't _quite_ line up with what was planned. But I kind of hope I can just keep writing and @ronny-mysten can update as needed :) 
- I'm adding @amnn because he had opinions about syntax in this doc and I don't remember what it was so hopefully he has some ideas.
- I'll ping people more broadly when the doc is complete

## Test Plan 

- 👀 

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes

amnn reviewed (commented):
Just did a brief run through -- on syntax, I think we ended up deciding that there were two documents that we wanted to write:

- one about how to create PTBs in the CLI that used some JSON-like syntax -- these are docs for the hypothetical feature in the CLI,
- another (this doc) about the semantics of PTB, which should use some language agnostic syntax for describing the relevant data structures.

I'm fine/happy with what it is like now, but we could probably avoid sticking exactly to the Rust syntax, by replacing `Vec<T>` with `[T]` and not necessarily needing to mention that `ObjectRef` is its own type, separate from the 3-tuple it represents, etc, but I don't feel strongly about this.

tnowacki commented on docs/content/concepts/transactions/WIP-ptb.md:
```diff
@@ -0,0 +1,101 @@
+
+In this document, we will be looking at the two parts of a programmable transaction block that are relevant to the exectuion semantics. Other transaction information, such as the transaction sender or the gas limit, might be referenced but are not out of scope. The programmable transaction block consists of two components
+
+- The inputs, `Vec<CallArg>`, is a vector of arguments, either objects or pure values, that can be used in the transaction commands. The objects are either owned by the sender or are shared/immutable objects. The pure values represent simple Move values, such as `u64` or `String` values, which can be constructed purely by their bytes.
```
I only want to reference a non builtin type here. I go into more detail for strings later (which might not have been there when you commented)

awelc commented on docs/content/concepts/transactions/WIP-ptb.md:
```diff
@@ -0,0 +1,400 @@
+#### Arguments
+
+- Each argument can be used by-reference or by-value. The usage is based on the type of the argument and the type signature of the command.
+  - If the signature expects an `&mut T`, the runtime checks the argument has type `T` and it is then mutably borrowed.
```
It feels like a lot to digest before getting to the description of commands, and I am not sure all of this needs to be digested before reading description of commands. Should we perhaps consider moving this to some kind of "additional details" section (at the end of "Executing a Command")?

tnowacki replied:
This document is supposed to be more of a reference, so I am fine leaning a bit heavier here. I had intended to put in all the detail here because it allows me to accurately describe each command more succinctly, though I totally get what you are saying. 

Maybe @ronny-mysten will have ideas?

ronny-mysten replied:
I'm of the same mind as Todd. I don't think it's an issue because I don't see it as a doc someone will read from beginning to end at once. We can always revisit based on user feedback.

awelc commented on docs/content/concepts/transactions/WIP-ptb.md:
```diff
@@ -0,0 +1,400 @@
+  - The `Item` with `id1` is a newly created object.
+- The item with `id2` is transferred to the sender's address.
+  - The `Item` with `id2` is a newly created object.
+- The `Marketplace` object is returned remains shared and its mutated.
```
```suggestion
- The `Marketplace` object is returned, remains shared, and it's mutated.
```
I am not sure if I got the intent of this sentence right, though...

tnowacki replied:
I think you did. Thanks!

awelc commented on docs/content/concepts/transactions/WIP-ptb.md:
```diff
@@ -0,0 +1,400 @@
+
+Programmable transaction blocks are used to define all user transactions on Sui. These transactions allow a user to call multiple Move functions, manage their objects, and manage their coins in a single transaction--without publishing a new Move package! Additionally, the structure of programmable transaction blocks was designed with automation and transaction builders in mind. In other words, th ...
+
+Each programmable transaction block is comprised of individual transaction commands (sometimes referred to themselves as transactions). Each command is executed in order, and the results from a command can be used in any subsequent command. The effects, i.e. object modifications or transfers, of all commands in a block are applied atomically at the end of the transaction, and if one command fails ...
```
```suggestion
Each programmable transaction block is comprised of individual transaction commands (sometimes referred to themselves as transactions or simply commands). Each command is executed in order, and the results from a command can be used in any subsequent command. The effects, i.e. object modifications or transfers, of all commands in a block are applied atomically at the end of the transaction, and if one command fails, the entire block fails and no effects from the commands are applied.
```

awelc commented on docs/content/concepts/transactions/prog-txn-blocks.mdx:
```diff
@@ -7,28 +7,228 @@ In basic terms, a blockchain is a block of transactions chained together in a sh
+On Sui, a transaction is more than a basic record of the flow of assets. Transactions on Sui are composed of a number of commands that execute on inputs to define the result of the transaction. Termed programmable transaction blocks (PTBs), these groups of commands define all user transactions on Sui. PTBs allow a user to call multiple Move functions, manage their objects, and manage their coins  ...
 
-:::info
+As mentioned, each PTB is comprised of individual transaction commands (sometimes referred to themselves as transactions). Each transaction command executes in order, and you can use the results from one transaction command in any subsequent transaction command. The effects, specifically object modifications or transfers, of all transaction commands in a block are applied atomically at the end of ...
```
```suggestion
As mentioned, each PTB is comprised of individual transaction commands (sometimes referred to themselves as transactions or simply commands). Each command executes in order, and you can use the results from one command in any subsequent transaction command. The effects, specifically object modifications or transfers, of all commands in a block are applied atomically at the end of the transaction. If one command fails, the entire block fails and no effects from the commands are applied.
```

For consistency with the other document (also, arguably reads a bit better without that many "transactions").

Reading further, though, I see that this document favors "transaction commands" over "commands". Perhaps I am too picky about it, but I feel like it would be better to be consistent in this.

awelc commented on docs/content/concepts/transactions/prog-txn-blocks.mdx:
```diff
@@ -7,28 +7,228 @@ In basic terms, a blockchain is a block of transactions chained together in a sh
+The inputs, `[CallArg]`, is a vector of arguments, either objects or pure values, that can be used in the transaction commands. The objects are either owned by the sender or are shared/immutable objects. The pure values represent simple Move values, such as `u64` or `String` values, which can be constructed purely by their bytes.
 
-There are multiple types of commands that you can use in a PTB, starting with implementations of common asset manipulation operations such as coin merging or splitting, and ending with the ability to call any on-chain Move function. You can chain typed outputs that your PTB commands receive and use them as typed inputs in subsequent commands in your PTB. With the ability to chain commands, you ca ...
+The commands, `[Command]`, is a vector of transaction commands. The possible commands are:
```
Getting to here, I am lost on what the difference is between the two docs in this PR...

ronny-mysten replied:
It's called plagiarism ;)

tnowacki commented on docs/content/concepts/transactions/WIP-ptb.md:
```diff
@@ -0,0 +1,400 @@
+  - `MergeCoins` merges multiple (1 or more) coins into a single coin. Any `sui::coin::Coin` objects can be merged, as long as they are all of the same type.
+  - `MakeMoveVec` creates a vector (potentially empty) of Move values. This is used primarily to construct vectors of Move values to be used as arguments to `MoveCall`.
+  - `MoveCall` invokes either an `entry` or a `public` Move function in a published package.
+  - `Publish` creates a new package and calls the `init` function of each module in the package.
```
@ronny-mysten, how crazy do we want to get with backlinks here? Like we could do it all over the place if we thought it was appropriate

ronny-mysten replied:
I think they get in the way if overused. I'd suggest using them once per page...so like the first mention of publishing/upgrading but nowhere else. I made a Jira to do a pass on the document to link appropriately.

stefan-mysten reviewed (commented):
Excellent doc @tnowacki. I left a few nits, hope that they can help improve the text.

dariorussi reviewed (commented):
Pretty extensive work, thank you! (or good night and good luck? 😉 )
putting an overall comment here so to avoid "polluting" the docs inline. All personal opinions here, do as you wish with it

- would it make sense to describe a PTB in terms of `[CallArg]` and `[Command]` in the "Transaction Type" paragraph? is that somewhere already? I read them but I have nothing to reference to and I am not sure about it...
- "Arguments and Results" wonder if a little (or not so little) picture of the 2D array and commands would speak volumes...
- are "Inputs" and "CallArgs" the same thing?
- "If an argument is copied, there must be no outstanding mutable borrows." - why?
- when talking about `sui::transfer::Receiving<T>` and shared object later would it make sense to have a reference to the related docs? (do they exist already?)
- `Publish` and `Upgrade` may benefit from a link to the publish and upgrade docs
- `Upgrade` in the general description of "Results" says there is no result and in the specification/explanation of the command it says "The command produces a single result type..."
- "Any remaining gas is returned to the gas coin, even if the owner has changed" - what does it mean?
- The final example looks good to me. I think that in the description of each command it may be more clear if the checks where in bullet points (like the end of execution) rather than in a "discursive" form. 

After reading it all I am more convinced that an initial "syntax" for a PTB would help tremendously (I know, I know, you said it first...). 
Possibly something not too formal and with comments, so people can just relate to the comment, but I am really convinced it would help the story overall a lot.

Finally I would suggest, after addressing the comments already there, to take a couple of days off and come back to this with a fresh mind. I think the docs would overall benefit from multiple passes by you (as long as they are with a fresh mind)

tnowacki commented:
> * would it make sense to describe a PTB in terms of `[CallArg]` and `[Command]` in the "Transaction Type" paragraph? is that somewhere already? I read them but I have nothing to reference to and I am not sure about it...

Done

> * are "Inputs" and "CallArgs" the same thing?

Fixed, I copied the Rust name for the type which really only exists for historical reasons 

> * "If an argument is copied, there must be no outstanding mutable borrows." - why?

Because I assumed I knew the rules without checking ;) 

> * when talking about `sui::transfer::Receiving<T>` and shared object later would it make sense to have a reference to the related docs? (do they exist already?)
> * `Publish` and `Upgrade` may benefit from a link to the publish and upgrade docs

I think @ronny-mysten has a task created to do this as a follow up

> * `Upgrade` in the general description of "Results" says there is no result and in the specification/explanation of the command it says "The command produces a single result type..."

Fixed

> * "Any remaining gas is returned to the gas coin, even if the owner has changed" - what does it mean?

I reworded, let me know if it helps

> * The final example looks good to me. I think that in the description of each command it may be more clear if the checks where in bullet points (like the end of execution) rather than in a "discursive" form.

Not exactly sure what you want here. Do you have an example?


> * "Arguments and Results" wonder if a little (or not so little) picture of the 2D array and commands would speak volumes...
> After reading it all I am more convinced that an initial "syntax" for a PTB would help tremendously (I know, I know, you said it first...). Possibly something not too formal and with comments, so people can just relate to the comment, but I am really convinced it would help the story overall a lot.
> Finally I would suggest, after addressing the comments already there, to take a couple of days off and come back to this with a fresh mind. I think the docs would overall benefit from multiple passes by you (as long as they are with a fresh mind)

I generally agree, and I think there is a bit of massaging to be done with the `Example` too. 
I think at this point, we should land what I have here, mostly for @ronny-mysten's sanity. And then I can come back next week or so and make another pass at it once I have sort of forgotten what I've written.
